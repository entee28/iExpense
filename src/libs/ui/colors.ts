export default {
  primary5: '#F7FAFF',
  primary10: '#E5EEFF',
  primary20: '#0055FF33',
  primary30: '#99BBFF',
  primary40: 'rgba(0, 85, 255, 0.4)',
  primary50: 'rgba(0, 85, 255, 0.5)',
  primary80: '#3377FF',
  primary100: '#0055FF',
  mono5: '#1A1A1A0D',
  mono4: '#1A1A1A0A',
  mono10: '#1A1A1A1A',
  mono40: '#1A1A1A66',
  mono60: '#1A1A1A99',
  mono70: '#1A1A1AB2',
  mono80: '#1A1A1ACC',
  mono100: '#1A1A1A',
  warning5: '#FFA8000D',
  warning10: '#FFA8001A',
  warning30: '#FFA8004D',
  warning80: '#FFA80080',
  warning100: '#FFA800',
  warning: '#FF6100',
  destructive: '#E60A32',
  destructive80: '#E60A32CC',
  destructive30: '#E60A324D',
  destructive30Hard: '#F8B6C1',
  destructive10: '#E60A321A',
  destructive5: 'rgba(230, 10, 50, 0.05)',
  positive5: '#00B47E0D',
  positive10: '#00B47E1A',
  positive30: '#00B47E4D',
  positive100: '#00B47E',
  white: '#FFFFFF',
  transparent: 'transparent',
  aliceBlue: '#EBF3FF'
}
