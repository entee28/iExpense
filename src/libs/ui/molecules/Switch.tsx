import React, { useEffect, useRef } from 'react'
import {
  Animated,
  Easing,
  StyleSheet,
  TouchableWithoutFeedback
} from 'react-native'
import colors from '../colors'

type Props = {
  value: boolean
  onChange?: (newValue: boolean) => void
  disabled?: boolean
  accessibilityLabel?: string
  trackTestID?: string
}

export const Switch = (props: Props) => {
  const { value, onChange, disabled, accessibilityLabel, trackTestID } = props

  const ballLeft = useRef(
    new Animated.Value(value ? MAX_BALL_LEFT : MIN_BALL_LEFT)
  ).current

  const handlePress = () => {
    onChange?.(!value)
  }

  useEffect(() => {
    Animated.timing(ballLeft, {
      toValue: value ? MAX_BALL_LEFT : MIN_BALL_LEFT,
      duration: ANIMATION_DURATION,
      useNativeDriver: true,
      easing: Easing.ease
    }).start()
  }, [value, ballLeft])

  const getTrackBackgroundColor = () => {
    if (value && disabled) {
      return colors.primary30
    }

    if (value && !disabled) {
      return colors.primary100
    }

    if (!value && disabled) {
      return colors.mono5
    }

    return colors.mono40
  }

  return (
    <TouchableWithoutFeedback
      testID={accessibilityLabel}
      onPress={disabled ? undefined : handlePress}>
      <Animated.View
        testID={trackTestID}
        style={[
          styles.track,
          {
            backgroundColor: getTrackBackgroundColor()
          }
        ]}>
        <Animated.View
          style={[
            styles.ball,
            {
              transform: [
                {
                  translateX: ballLeft
                }
              ]
            }
          ]}
        />
      </Animated.View>
    </TouchableWithoutFeedback>
  )
}

const ANIMATION_DURATION = 150
const BALL_SIZE = 20
const TRACK_PADDING = 2
const TRACK_HEIGHT = BALL_SIZE + TRACK_PADDING * 2
const TRACK_WIDTH = 40
export const MIN_BALL_LEFT = 0
export const MAX_BALL_LEFT = TRACK_WIDTH - BALL_SIZE - TRACK_PADDING * 2

const styles = StyleSheet.create({
  track: {
    width: TRACK_WIDTH,
    height: TRACK_HEIGHT,
    borderRadius: 9999,
    padding: TRACK_PADDING
  },
  ball: {
    width: BALL_SIZE,
    height: BALL_SIZE,
    borderRadius: 9999,
    backgroundColor: colors.white
  }
})
