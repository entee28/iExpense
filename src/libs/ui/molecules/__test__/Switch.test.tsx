import { fireEvent, render } from '@testing-library/react-native'
import colors from '../../colors'
import { Switch } from '../Switch'

it('should toggle value on press', () => {
  let value = false
  const setValue = (newValue: boolean) => {
    value = newValue
  }

  const testSwitch = render(
    <Switch value={value} onChange={setValue} accessibilityLabel="switch" />
  )

  fireEvent.press(testSwitch.getByTestId('switch'))
  expect(value).toEqual(true)
})

it('should renders state correctly', () => {
  const offSwitch = render(<Switch value={false} trackTestID="track" />)
  const offDisabledSwitch = render(
    <Switch value={false} disabled trackTestID="track" />
  )
  const onSwitch = render(<Switch value={true} trackTestID="track" />)
  const onDisabledSwitch = render(
    <Switch value={true} disabled trackTestID="track" />
  )

  expect(offSwitch.getByTestId('track')).toHaveStyle({
    backgroundColor: colors.mono40
  })

  expect(offDisabledSwitch.getByTestId('track')).toHaveStyle({
    backgroundColor: colors.mono5
  })

  expect(onSwitch.getByTestId('track')).toHaveStyle({
    backgroundColor: colors.primary100
  })

  expect(onDisabledSwitch.getByTestId('track')).toHaveStyle({
    backgroundColor: colors.primary30
  })
})

it('should match snapshot', () => {
  let value = false
  const setValue = (newValue: boolean) => {
    value = newValue
  }
  const snapshot = render(
    <Switch
      value={value}
      onChange={setValue}
      accessibilityLabel="switch"
      disabled={false}
    />
  ).toJSON()

  expect(snapshot).toMatchSnapshot()
})
