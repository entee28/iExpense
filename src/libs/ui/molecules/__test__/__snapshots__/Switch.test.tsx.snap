// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#1A1A1A66",
      "borderRadius": 9999,
      "height": 24,
      "padding": 2,
      "width": 40,
    }
  }
  testID="switch"
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 9999,
        "height": 20,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": 20,
      }
    }
  />
</View>
`;
